Update e2e scripts to run storage tests on GCE/Vagrant

jsafrane opened the pull request:
In order to run the tests/e2e/volumes.go on GCE or Vagrant, this patch introduces `KUBE_E2E_TEST_ENVIRONMENT` variable. It's automatically set to `true` by `go run hack/e2e.go --up`. When set to `true`, it:
- Installs the right packages for all storage volumes (mount.gluster, rbd, iscsiadm, ...)
- Configures docker to use devicemapper as docker storage backend. 'aufs', the default one on Debian, does not support extended attributes, required by Ceph RBD and Gluster server containers.
- Enable '--allow-privileged=True' in kube-apiserver and kubelet.

Clusters created by `cluster/kube-up.sh` without setting `KUBE_E2E_TEST_ENVIRONMENT` are not affected and cannot run the volume tests.

Tested on GCE and Vagrant, together with PR #11054.

~~Also, the second patch fixes a busy loop in the containers.~~

satnam6502 commented:
@zmerlynn

ixdy commented on cluster/aws/config-default.sh:
```diff
@@ -94,3 +94,5 @@ COREOS_CHANNEL="${COREOS_CHANNEL:-alpha}"
 RKT_VERSION="${KUBE_RKT_VERSION:-0.5.5}"
 
+# Optional: if set to true, kube-up will configure the cluster to run e2e tests.
+E2E_TEST_ENVIRONMENT=${KUBE_E2E_TEST_ENVIRONMENT:-false}
```
following the convention of other boolean environment variables, this should probably be something like `ENABLE_E2E_TEST_ENVIRONMENT`.

jsafrane replied:
Renamed

ixdy commented on cluster/saltbase/salt/docker/docker-defaults:
```diff
@@ -1,3 +1,7 @@
-DOCKER_OPTS="{{grains_opts}} --bridge=cbr0 --iptables=false --ip-masq=false"
+{% set e2e_opts = '' -%}
+{% if pillar.get('e2e_test_environment', '').lower() == 'true' -%}
+{% set e2e_opts = '-s devicemapper' -%}
```
I'm a little worried about using a different docker driver in tests. It seems like this would be likely to mask bugs that'd only be seen in production.

roberthbailey replied:
+1

jsafrane replied:
Extended attributes (=`devicemapper` storage driver)  are necessary to run Gluster and Ceph containers. I guess I could create a 1GB file *inside* the containers, format it as ext3 and create Gluster/Ceph bricks there, but it looks like an overkill to me.

IMO it's not up to Kubernetes to decide what storage backend the system admin configures for Docker, it should work with all of them. Just for e2e tests it must be something that supports extended attributes inside the containers.

derekwaynecarr replied:
I agree with @jsafrane that it should not be a Kubernetes decision point what storage driver is used for Docker.

derekwaynecarr replied:
that said, it seems like coupling e2e_opts to the storage driver is a little too subtle.

can we just have a separate flag in salt that controls devicemapper only?  then have a config-e2e.sh that uses that option if required?

ixdy commented:
cc @dchen1107 since I'm not super familiar with salt stuff

jsafrane commented:
Rebased to today;s master.

ixdy commented:
@zmerlynn @roberthbailey would either of you be able to look at this? I don't feel knowledgeable enough in any of the affected code.

roberthbailey commented on cluster/saltbase/salt/kubelet/default:
```diff
@@ -102,5 +102,10 @@
 
+{% set privileged = pillar.get('allow_privileged', false)  -%}
+{% if pillar.get('e2e_test_environment', '').lower() == 'true' -%}
+  {% set privileged = "True" -%}
```
I'm not a huge fan of turning on privileged by default for all e2e runs. We are likely to end up in a situation where we inadvertently introduce a dependency on privileged mode that can't be detected by our test infrastructure.

zmerlynn replied:
+1

zmerlynn replied:
I think we're about to get into a divide on "a set of tests that can run on more or less vanilla clusters" and "stuff that needs invasive permissions". Can you really not do this without such invasive changes? I'm not totally against the package changes (maybe) but the environmental changes are going to block this for now.

jsafrane replied:
> Can you really not do this without such invasive changes?

No, PersistentVolume, NFS, Gluster, Ceph, and iSCSI tests need `--allow-privileged=True` to work. It seems it's really necessary to have two separate test environments. I can rename `ENABLE_E2E_TEST_ENVIRONMENT` to anything else, still, it will probably result in a new Jenkins environment.

roberthbailey commented:
If we want to make a bunch of customizations only for the purposes of running these tests, I'd suggest that we leave the default e2e configuration as it is and introduce a new configuration that changes docker / enables privileged mode just to run these tests. Then we can work to reduce the differences between the special tests and the normal tests and add these tests to the standard suite once the differences have been eliminated.

jsafrane commented:
I am not sure I understand. Do you want me to rename KUBE_ENABLE_E2E_TEST_ENVIRONMENT to say... KUBE_ENABLE_STORAGE_TEST_ENVIRONMENT (+ probably rename e2e/init.sls to something storage-test-related)

jsafrane commented:
The test failed with:
> • Failure [110.564 seconds]
> Services
> /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:946
>   should serve multiport endpoints from pods [It]
>   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:235

>   Aug 24 00:59:35.527: Timed out waiting for service multi-endpoint-test in namespace e2e-tests-service-0-v2h96 to expose endpoints map[podname1:[100]] (1m0s elapsed)

I don't think it's caused by this PR, it must be a flaky test.

zmerlynn commented on cluster/gce/util.sh:
```diff
@@ -574,6 +574,7 @@ function create-certs {
 #   KUBE_ROOT
 #   <Various vars set in config file>
 function kube-up {
+  set -x
```
Debug?

jsafrane replied:
good catch, fixed

zmerlynn commented:
Apologies for not reviewing this much earlier, but yes, I'm in agreement with other reviewers. This is basically a separate environment, and we would essentially have to run a separate Jenkins job to avoid polluting other E2E runs, because it's *not* a general E2E environment.

I'll be on vacation for a week starting tomorrow. Again, apologies for the late review. I can glance at it when I get back. The Salt looked mostly okay, I'm just philosophically against polluting the general e2e environment.

jsafrane commented:
So... how to proceed with this PR? These changes are necessary for volume e2e tests. I understand that they are not welcome for 'regular' testing. Should there be two sets of tests? How it should look like?

zmerlynn commented:
Now that #13930 is in, it seems like the story is a little different, I presume?

zmerlynn commented:
Ping? I think this PR is different on 1.1, no with privileged mode enabled by default?

Actually, are the docker options still necessary on Docker 1.8.3, or did that part evaporate, too?

zmerlynn commented:
Reassign back when you're ready.

jsafrane commented:
Sorry, updating salt scripts is quite time-consuming and I always got distracted by something more important... Now it's rebased and I removed handling of privileged mode.

> Actually, are the docker options still necessary on Docker 1.8.3, or did that part evaporate, too?

Docker 1.8.3 does not help, it still runs with aufs backend on Debian. The tests don't work with it and they still need devicemapper backend.

jsafrane commented:
@zmerlynn, it seems I cannot re-assign people, poking this way.

zmerlynn commented:
We still haven't figured out the general environmental question. This creates a really heavy environmental difference for the e2e tests, but should probably be a twiddle for the storage tests alone (or ideally some subset of runs).

cc @ixdy @thockin @quinton-hoole, this has been lagging a while, let's see if we can try to come to a consensus on how to get this in.

quinton-hoole commented on cluster/azure/templates/common.sh:
```diff
@@ -0,0 +1,58 @@
+  local -r url="$1"
+  local -r file="${url##*/}"
+  rm -f "$file"
+  until [[ -e "${file}" ]]; do
```
This is potentially an endless loop?  Add timeout or max_retries?

jsafrane replied:
Wait, the patch should not aggressively modify Azure, something got wrong when rebasing... I'll post a new patch shortly. (btw, good catch!)

jsafrane replied:
Indeed, Azure was removed by https://github.com/kubernetes/kubernetes/pull/15790, I updated the patch.

quinton-hoole commented on cluster/azure/templates/common.sh:
```diff
@@ -0,0 +1,58 @@
+# Retry a download until we get it.
+#
+# $1 is the URL to download
+download-or-bust() {
```
Prefer less ambiguous name.  "download-with-retires"?

quinton-hoole replied:
Would prefer if this took the output file as an explicit argument, rather than implying it from the url.

quinton-hoole commented:
@jsafrane @zmerlynn Yes I agree with the other reviewers.  The custom environment needs to be specific to the set of storage tests that require it (with names reflecting that fact as per @jsafrane 's comment of Aug 19, and we'll need a separate CI test job that invokes this custom environment.

As far as I can see, we're all in agreement on the above, so I declare this consensus unless anyone objects :-)

zmerlynn commented on cluster/saltbase/salt/e2e/init.sls:
```diff
@@ -0,0 +1,41 @@
+      - open-iscsi
+      - iscsitarget-dkms
+      - nfs-common
+      - nfs-kernel-server
```
I'm not sure you need this on the base image. At least, I ran through the entire NFS example without needing this package installed on `containervm` nodes.

jsafrane replied:
Indeed, `nfs-kernel-server` is not needed. I removed it.

zmerlynn commented on cluster/saltbase/salt/e2e/init.sls:
```diff
@@ -0,0 +1,41 @@
+
+  # Install various packages required by e2e tests to all hosts.
+  # 'pkg.latest' is used to install updated glusterfs-client from the repo above,
+  # GCE image already has glusterfs-client 3.2.5, which is too old.
```
I'll try to fix that next CVM build. Thanks for the heads up.

zmerlynn commented:
@jsafrane: Judging by how this your environment variable works, I think you could actually get away with just renaming this `KUBE_E2E_STORAGE_TEST_ENVIRONMENT`. What do you think, @quinton-hoole?

That basically knocks this out of GKE testing, which is unfortunate. We won't be able to support this uber-unique, special snowflake bringup, and I have a *lot* of concerns about this environment.

Is there a way you can also attack this on the e2e side by interrogating whether the cluster is capable, rather than beating the cluster into being capable? Then I can push GKE there instead.

quinton-hoole commented on cluster/gce/config-test.sh:
```diff
@@ -135,3 +135,6 @@ NETWORK_PROVIDER="${NETWORK_PROVIDER:-none}" # opencontrail
 OPENCONTRAIL_PUBLIC_SUBNET="${OPENCONTRAIL_PUBLIC_SUBNET:-10.1.0.0/16}"
+
+# Optional: if set to true, kube-up will configure the cluster to run e2e tests.
+ENABLE_E2E_TEST_ENVIRONMENT=${KUBE_ENABLE_E2E_TEST_ENVIRONMENT:-true}
```
As per other review comments, the default should be false, even for test evironments, and should need to be explicitly overridden for specialised storage test environments.

jsafrane replied:
E2E_STORAGE_TEST_ENVIRONMENT (was ENABLE_E2E_TEST_ENVIRONMENT) is now `false` by default in all scripts.

quinton-hoole commented on cluster/vagrant/config-test.sh:
```diff
@@ -23,3 +23,6 @@ source "${KUBE_ROOT}/cluster/vagrant/config-default.sh"
 REGISTER_MASTER_KUBELET=${REGISTER_MASTER:-false}
+
+# Optional: if set to true, kube-up will configure the cluster to run e2e tests.
+ENABLE_E2E_TEST_ENVIRONMENT=${KUBE_ENABLE_E2E_TEST_ENVIRONMENT:-true}
```
Same comment as above.  Default to false.

quinton-hoole commented on cluster/vsphere/config-test.sh:
```diff
@@ -32,3 +32,6 @@ MINION_MEMORY_MB=1024
 SERVICE_CLUSTER_IP_RANGE="10.244.240.0/20"  # formerly PORTAL_NET
+
+# Optional: if set to true, kube-up will configure the cluster to run e2e tests.
+ENABLE_E2E_TEST_ENVIRONMENT=${KUBE_ENABLE_E2E_TEST_ENVIRONMENT:-true}
```
Same comment as above.  Default to false.

quinton-hoole commented:
@zmerlynn @jsafrane Yes, I'm happy with the simple solution proposed by @zmerlynn above.  See also my immediately recent in-code comments.

jsafrane commented:
Updated:
* renamed KUBE_E2E_TEST_ENVIRONMENT to E2E_STORAGE_TEST_ENVIRONMENT
* is `false` by default in all scripts
* removed a nfs server package from Debian setup

zmerlynn commented:
LGTM. Looks like you got hit with another rebase, so I haven't bothered labelling it. Rebase and I'll give it a quick once over and we'll be ready to go. cc @quinton-hoole if you want another pass.

quinton-hoole commented:
@zmerlynn LGTM

zmerlynn commented:
Gratuitously labeling. @k8s-merge-robot will take it off after you update, but it'll remind me to relabel.

zmerlynn commented:
@k8s-oncall: @k8s-merge-robot lost its brain

saad-ali commented:
I can merge this manually. Can you rebase?

zmerlynn commented on cluster/saltbase/salt/e2e/init.sls:
```diff
@@ -0,0 +1,40 @@
+{% else %}
+      - glusterfs-client
+      - open-iscsi
+      - iscsitarget-dkms
```
Ack, sorry. I was updating the ContainerVM manifest and if I'm reading this module correctly, this is the source code module. Do you need `iscsitarget-dkms` or just `open-iscsi`?

zmerlynn replied:
The weird thing is, I *think* you want `iscsitarget` instead, and maybe `iscsitarget-module`, but the latter is a virtual target that doesn't seem to resolve. :/

zmerlynn replied:
Ah, I just followed through the `iscsi` example and understand now. That's .. odd, but okay.

jsafrane commented:
Rebased

zmerlynn commented:
LGTM

zmerlynn commented:
@k8s-bot test this

beni05 commented:
úgy tűnik, hogy kimaradt, zárás ,!

zmerlynn commented:
Hrm. @k8s-bot really doesn't want to tag back in (just checked it internally). I'm going to try close/reopen as well.

zmerlynn commented:
cc @ixdy

Removing LGTM to stop spam, but not sure why @k8s-bot won't come in

zmerlynn commented:
@jsafrane If you rebase this, I'll manually merge on green

jsafrane committed:
Configure cluster for e2e tests.

When KUBE_E2E_STORAGE_TEST_ENVIRONMENT is set to 'true', kube-up.sh script
will:

- Install the right packages for all storage volumes.
- Use devicemapper as docker storage backend. 'aufs', the default one on
Debian, does not support extended attibutes required by Ceph RBD and Gluster
server containers.

Tested on GCE and Vagrant, e2e tests for storage volumes passes without any
additional configuration.

jsafrane commented:
Rebased

zmerlynn commented:
@k8s-bot unit test this

Still LGTM, will watch for green.

zmerlynn commented:
@k8s-bot test this

zmerlynn commented:
Jenkins is about to report green, and it looks like the SQ has caught up. If this isn't merged in a couple of hours, I'll hit the button. (If it needs a rebase again *already*, I'll eat my words and owe you so many drinks-of-choice.)

zmerlynn commented:
@k8s-bot test this

wojtek-t commented:
Jenkins failure seems like an infrastructural problem:

    06:55:31 Creating new network: pull-e2e-parallel-2
    06:55:34 ERROR: (gcloud.compute.networks.create) Some requests did not succeed:
    06:55:34  - The resource 'projects/kubernetes-jenkins-pull/global/networks/pull-e2e-parallel-2' already exists
    06:55:34 
    06:55:35 2015/10/29 06:55:35 e2e.go:309: Error running up: exit status 1
    06:55:35 2015/10/29 06:55:35 e2e.go:305: Step 'up' finished in 1m6.204835232s
    06:55:35 2015/10/29 06:55:35 e2e.go:141: Error starting e2e cluster. Aborting.

@ixdy - FYI ^

zmerlynn commented:
@wojtek-t: Yeah, as long as I hit the right workspace, it works. :/

zmerlynn commented:
@wojtek-t: Not even sure it's the workspace. Got past it this time, certainly, but looking at the code, there's no reason for that to fail that way. (We once or twice had a `PROJECT` mismatch bug, but that's gone.) It almost looks like a `gcloud` flake, like the exit code was wrong, but that feels like I'm claiming sunspots.

zmerlynn commented:
I'm manually merging given PR builder flakiness.

jsafrane commented:
Hurray, thanks everyone for review & merge!
